src/gwt/src/org/rstudio/studio/client/workbench/model/SessionOpener.java

```diff
@@ -211,13 +213,32 @@ public boolean execute()
          {
             // if we've already delivered the ping or our retry count
             // is exhausted then return false
-            if (pingDelivered_ || (++retries_ > maxRetries))
+            if (pingDelivered_)
+            {
                return false;
+            }
+            
+            // if we hit our retry count, give up
+            if (retries_++ > maxRetries)
+            {
+               Debug.logWarning("Error connecting with session.");
+               return false;
+            }
+            
+            // if we have a ping in flight, but we don't receive a response
```

Comment:
I made this change because I noticed, on debug builds, when restarting the R session, we seemingly could send off a "ping" request but never actually receive a response -- as though the ping RPC was either never sent, or never processed by the new session.

The downside here is we might end up with an extra ping or two in-flight while a slow session restarts, but I think that's okay?